File: typescript/packages/core/src/core/jq/stream.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { AsyncLineIterator } from '../../io/async_line_iterator.ts'
import { formatOne } from './format.ts'
import { RS, type JqOptions } from './types.ts'

const DEC = new TextDecoder('utf-8', { fatal: false })
const WHITESPACE = /\s/

function parseSafe(text: string): unknown {
  return JSON.parse(text) as unknown
}

function parseJsonl(raw: Uint8Array): unknown[] {
  const text = DEC.decode(raw)
  return text
    .split('\n')
    .filter((line) => line.trim() !== '')
    .map((line) => parseSafe(line))
}

function documentEnd(text: string, start: number): number {
  const opener = text[start]
  if (opener !== '{' && opener !== '[') {
    if (opener !== '"') {
      let i = start
      while (i < text.length) {
        const ch = text[i]
        if (ch === undefined || WHITESPACE.test(ch)) break
        i += 1
      }
      return i
    }
    let i = start + 1
    while (i < text.length) {
      const ch = text[i]
      if (ch === '\\') {
        i += 2
        continue
      }
      if (ch === '"') return i + 1
      i += 1
    }
    return text.length
  }
  let depth = 0
  let inStr = false
  let i = start
  while (i < text.length) {
    const ch = text[i]
    if (inStr) {
      if (ch === '\\') i += 1
      else if (ch === '"') inStr = false
      i += 1
      continue
    }
    if (ch === '"') inStr = true
    else if (ch === '{' || ch === '[') depth += 1
    else if (ch === '}' || ch === ']') {
      depth -= 1
      if (depth === 0) return i + 1
    }
    i += 1
  }
  return text.length
}

/**
 * Parse a whitespace-separated stream of JSON values.
 *
 * Empty input holds no documents at all, which is why jq prints nothing
 * and exits 0 for an empty file.
 */
export function parseJsonDocs(raw: Uint8Array): unknown[] {
  const text = DEC.decode(raw).trim()
  if (text === '') return []
  try {
    return [parseSafe(text)]
  } catch (singleDocError) {
    const docs: unknown[] = []
    let idx = 0
    while (idx < text.length) {
      const end = documentEnd(text, idx)
      if (end <= idx) throw singleDocError
      try {
        docs.push(parseSafe(text.slice(idx, end)))
      } catch {
        // Not a value stream either, so the input is simply invalid.
        // Re-throw the whole-document error: it names the real problem.
        throw singleDocError
      }
      idx = end
      while (idx < text.length) {
        const ch = text[idx]
        if (ch === undefined || !WHITESPACE.test(ch)) break
        idx += 1
      }
    }
    return docs
  }
}

export function parseJsonAuto(raw: Uint8Array): unknown {
  const docs = parseJsonDocs(raw)
  if (docs.length === 0) throw new Error('jq: empty input')
  return docs.length === 1 ? docs[0] : docs
}

/**
 * Parse an RFC 7464 JSON text sequence (`--seq`).
 *
 * Every value is introduced by RS, so anything before the first one is
 * text the sequence never claimed. jq reports that as an ignored parse
 * error and prints nothing for it; mirage drops it just as silently,
 * which is the one divergence here.
 */
export function parseSeqDocs(raw: Uint8Array): unknown[] {
  return DEC.decode(raw)
    .split(RS)
    .slice(1)
    .filter((part) => part.trim() !== '')
    .map((part) => parseSafe(part))
}

/**
 * Split one input into the strings `jq -R` reads it as.
 *
 * jq breaks on newlines only (never on the other separators a Unicode
 * line splitter honors) and a trailing newline ends the last line rather
 * than starting an empty one.
 */
export function splitRawLines(raw: Uint8Array): string[] {
  const text = DEC.decode(raw)
  if (text === '') return []
  const lines = text.split('\n')
  if (lines[lines.length - 1] === '') lines.pop()
  return lines
}

export function parseJsonPath(raw: Uint8Array, path: string): unknown {
  if (path.endsWith('.jsonl') || path.endsWith('.ndjson')) return parseJsonl(raw)
  return parseSafe(DEC.decode(raw))
}

export function isJsonlPath(path: string): boolean {
  return path.endsWith('.jsonl') || path.endsWith('.ndjson')
}

export function isStreamableJsonlExpr(expression: string): boolean {
  return expression.trim().startsWith('.[]')
}

/**
 * Evaluate a per-element program over a JSONL file, line by line.
 *
 * Only output options reach here, since the caller keeps this path off
 * for anything that changes input assembly.
 */
export async function* evalJsonlStream(
  source: AsyncIterable<Uint8Array>,
  expression: string,
  opts: JqOptions,
): AsyncIterable<Uint8Array> {
  const { argsObject, jqEval, referencesArgs } = await import('./eval.ts')
  const expr = expression.trim()
  let perItem: string
  if (expr === '.[]') perItem = '.'
  else if (expr.startsWith('.[] | ')) perItem = expr.slice(6)
  else if (expr.startsWith('.[].')) perItem = expr.slice(3)
  else perItem = expr

  const argsValue = referencesArgs(perItem) ? argsObject(opts) : null
  const iter = new AsyncLineIterator(source)
  for await (const lineBytes of iter) {
    const text = DEC.decode(lineBytes).trim()
    if (text === '') continue
    const obj: unknown = JSON.parse(text)
    for (const value of await jqEval(obj, perItem, opts.namedArgs, null, argsValue)) {
      yield formatOne(value, opts)
    }
  }
}

File: python/mirage/core/jq/stream.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import json
from collections.abc import AsyncIterator
from typing import Any

import orjson

from mirage.core.jq.eval import args_object, jq_eval, references_args
from mirage.core.jq.format import format_one
from mirage.core.jq.types import RS, JqOptions
from mirage.io.async_line_iterator import AsyncLineIterator


def parse_jsonl(raw: bytes) -> list[Any]:
    text = raw.decode("utf-8", errors="replace")
    return [orjson.loads(line) for line in text.splitlines() if line.strip()]


def parse_json_docs(raw: bytes) -> list[Any]:
    """Parse a whitespace-separated stream of JSON values.

    jq reads a stream of values from any input, not one document: `.json`
    holding several values, newline-delimited JSONL, and pretty-printed
    values run together are all valid and all evaluate per document. The
    single-value case takes the fast orjson path; only a stream falls
    back to incremental decoding.

    Args:
        raw (bytes): the whole input.

    Returns:
        list[Any]: every decoded document, in order. Empty input holds
            no documents at all, which is why jq prints nothing and
            exits 0 for an empty file.
    """
    text = raw.decode("utf-8", errors="replace").strip()
    if not text:
        return []
    try:
        return [orjson.loads(text)]
    except orjson.JSONDecodeError as single_doc_error:
        first_error = single_doc_error
    decoder = json.JSONDecoder()
    docs: list[Any] = []
    idx = 0
    end = len(text)
    while idx < end:
        try:
            doc, offset = decoder.raw_decode(text, idx)
        except ValueError:
            # Not a value stream either, so the input is simply invalid.
            # Re-raise the whole-document error: it is the one that names
            # the real problem, and callers match on orjson's type.
            raise first_error from None
        docs.append(doc)
        idx = offset
        while idx < end and text[idx].isspace():
            idx += 1
    return docs


def parse_json_auto(raw: bytes) -> object:
    docs = parse_json_docs(raw)
    if not docs:
        raise ValueError("jq: empty input")
    return docs[0] if len(docs) == 1 else docs


def parse_seq_docs(raw: bytes) -> list[Any]:
    """Parse an RFC 7464 JSON text sequence (`--seq`).

    Every value is introduced by RS, so anything before the first one is
    text the sequence never claimed. jq reports that as an ignored parse
    error and prints nothing for it; mirage drops it just as silently,
    which is the one divergence here.

    Args:
        raw (bytes): the whole input.
    """
    text = raw.decode("utf-8", errors="replace")
    return [orjson.loads(part) for part in text.split(RS)[1:] if part.strip()]


def split_raw_lines(raw: bytes) -> list[str]:
    """Split one input into the strings `jq -R` reads it as.

    jq breaks on newlines only (never on the other separators
    ``str.splitlines`` honors) and a trailing newline ends the last line
    rather than starting an empty one.

    Args:
        raw (bytes): one input's bytes.
    """
    text = raw.decode("utf-8", errors="replace")
    if not text:
        return []
    lines = text.split("\n")
    if lines[-1] == "":
        lines.pop()
    return lines


def parse_json_path(raw: bytes, path: str) -> object:
    if path.endswith(".jsonl") or path.endswith(".ndjson"):
        return parse_jsonl(raw)
    return orjson.loads(raw)


def is_jsonl_path(path: str) -> bool:
    return path.endswith(".jsonl") or path.endswith(".ndjson")


def is_streamable_jsonl_expr(expression: str) -> bool:
    expr = expression.strip()
    if expr.startswith(".[]"):
        return True
    return False


async def eval_jsonl_stream(
    source: AsyncIterator[bytes],
    expression: str,
    opts: JqOptions,
) -> AsyncIterator[bytes]:
    """Evaluate a per-element program over a JSONL file, line by line.

    Args:
        source (AsyncIterator[bytes]): the file's byte chunks.
        expression (str): jq program text, already known to be one of
            the `.[]`-prefixed shapes this path can rewrite.
        opts (JqOptions): resolved options; only output ones reach here,
            since the caller keeps this path off for anything that
            changes input assembly.
    """
    expr = expression.strip()
    if expr == ".[]":
        per_item = "."
    elif expr.startswith(".[] | "):
        per_item = expr[6:]
    elif expr.startswith(".[]."):
        per_item = expr[3:]
    else:
        per_item = expr

    args_value = args_object(opts) if references_args(per_item) else None
    async for line_bytes in AsyncLineIterator(source):
        text = line_bytes.decode("utf-8", errors="replace").strip()
        if not text:
            continue
        obj = orjson.loads(text)
        for value in jq_eval(obj, per_item, opts.named_args, None, args_value):
            yield format_one(value, opts)
